{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d09f5859-55db-4cb0-9d86-b0909bacab00",
   "metadata": {},
   "source": [
    "# Ranges"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c12970f-4764-4628-b97d-0bd484a14b49",
   "metadata": {},
   "source": [
    "A [bioframe](https://bioframe.readthedocs.io/en/latest/) based realization of genomic ranges integration in MuData\n",
    "\n",
    "`RangeAnnData` is an extension class to the AnnData class"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82b63cef-bdc2-435e-99a3-59974e66ad78",
   "metadata": {},
   "source": [
    "## Prepare data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8b8c0d85-b2d2-4842-9c22-8987494199be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The history saving thread hit an unexpected error (OperationalError('attempt to write a readonly database')).History will not be written to the database.\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "from scipy.sparse import csr_matrix\n",
    "from mudata import MuData\n",
    "import mudata as md\n",
    "from anndata import AnnData\n",
    "import pyranges as pr\n",
    "import bioframe as bf\n",
    "\n",
    "from RangeAnnData import RangeAnnData, RangeMuData"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4303d432-1be7-4e92-9b26-d7011ecf73aa",
   "metadata": {},
   "source": [
    "## Generation of toy dataset\n",
    "\n",
    "- Count matrix: sparse data with `scipy.csr_matrix`\n",
    "- genomic ranges: provided by `PyRanges`\n",
    "- Make sure the naming of columns follow the bed file format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "87b7d40c",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = csr_matrix(np.random.poisson(1, size=(100, 1000)), dtype=np.float32)\n",
    "exons, gr = pr.data.exons().df, pr.data.cpg().df\n",
    "exons = exons.sample(frac=1, replace=False)\n",
    "\n",
    "exons.columns = ['chrom', 'start', 'end'] + list(exons.columns[3:])\n",
    "gr.columns = ['chrom', 'start', 'end']+ list(gr.columns[3:])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08d8e1b0",
   "metadata": {},
   "source": [
    "## Build AnnData object out of the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5423509d",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = RangeAnnData(counts)\n",
    "adata.obs_names = [f\"Cell_{i:d}\" for i in range(adata.n_obs)]\n",
    "adata.var_names = [f\"Gene_{i:d}\" for i in range(adata.n_vars)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a8ba350",
   "metadata": {},
   "source": [
    "Coordinates are integrated into `RangedAnnData` with `varm` by method `set_coord`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8952171c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gene_0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>147733519</td>\n",
       "      <td>147733640</td>\n",
       "      <td>NM_001169125_exon_1_0_chrX_147733520_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>30718954</td>\n",
       "      <td>30719022</td>\n",
       "      <td>NM_001128127_exon_9_0_chrX_30718955_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>16875716</td>\n",
       "      <td>16875832</td>\n",
       "      <td>NM_002893_exon_7_0_chrX_16875717_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>128674251</td>\n",
       "      <td>128674455</td>\n",
       "      <td>NM_000276_exon_0_0_chrX_128674252_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>83588766</td>\n",
       "      <td>83588850</td>\n",
       "      <td>NM_001177479_exon_2_0_chrX_83588767_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_995</th>\n",
       "      <td>chrX</td>\n",
       "      <td>15838329</td>\n",
       "      <td>15838439</td>\n",
       "      <td>NM_005089_exon_9_0_chrX_15838330_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_996</th>\n",
       "      <td>chrX</td>\n",
       "      <td>1407411</td>\n",
       "      <td>1407535</td>\n",
       "      <td>NM_172249_exon_4_0_chrX_1407412_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_997</th>\n",
       "      <td>chrX</td>\n",
       "      <td>31893307</td>\n",
       "      <td>31893490</td>\n",
       "      <td>NM_004010_exon_31_0_chrX_31893308_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_998</th>\n",
       "      <td>chrX</td>\n",
       "      <td>73744193</td>\n",
       "      <td>73744644</td>\n",
       "      <td>NM_006517_exon_2_0_chrX_73744194_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_999</th>\n",
       "      <td>chrX</td>\n",
       "      <td>101095987</td>\n",
       "      <td>101096076</td>\n",
       "      <td>NM_032946_exon_8_0_chrX_101095988_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chrom      start        end                                    Name  \\\n",
       "Gene_0    chrX  147733519  147733640  NM_001169125_exon_1_0_chrX_147733520_f   \n",
       "Gene_1    chrX   30718954   30719022   NM_001128127_exon_9_0_chrX_30718955_f   \n",
       "Gene_2    chrX   16875716   16875832      NM_002893_exon_7_0_chrX_16875717_r   \n",
       "Gene_3    chrX  128674251  128674455     NM_000276_exon_0_0_chrX_128674252_f   \n",
       "Gene_4    chrX   83588766   83588850   NM_001177479_exon_2_0_chrX_83588767_r   \n",
       "...        ...        ...        ...                                     ...   \n",
       "Gene_995  chrX   15838329   15838439      NM_005089_exon_9_0_chrX_15838330_f   \n",
       "Gene_996  chrX    1407411    1407535       NM_172249_exon_4_0_chrX_1407412_f   \n",
       "Gene_997  chrX   31893307   31893490     NM_004010_exon_31_0_chrX_31893308_r   \n",
       "Gene_998  chrX   73744193   73744644      NM_006517_exon_2_0_chrX_73744194_f   \n",
       "Gene_999  chrX  101095987  101096076     NM_032946_exon_8_0_chrX_101095988_r   \n",
       "\n",
       "          Score Strand  \n",
       "Gene_0        0      +  \n",
       "Gene_1        0      +  \n",
       "Gene_2        0      -  \n",
       "Gene_3        0      +  \n",
       "Gene_4        0      -  \n",
       "...         ...    ...  \n",
       "Gene_995      0      +  \n",
       "Gene_996      0      +  \n",
       "Gene_997      0      -  \n",
       "Gene_998      0      +  \n",
       "Gene_999      0      -  \n",
       "\n",
       "[1000 rows x 6 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.set_coord(exons)\n",
    "adata.varm['coord']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55275a34-6be8-4a98-b04d-c41bb53aa26d",
   "metadata": {},
   "source": [
    "## Filter by genomic ranges: `slice_granges`\n",
    "\n",
    "Original data size:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3cd52ab6-63fc-44ed-809d-2e5c6f604f7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(100, 1000)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f8513cf-9c31-4a45-b731-f55757607953",
   "metadata": {},
   "source": [
    "Sometimes features are defined in some coordinate system such as a linear _genome sequence_ — for instance in assays that measure chromatin accessibility, transcription factor occupancy, DNA methylation. In fact, genes also have a property defining their location in the DNA, even though that's something frequently ignored in transcriptomics analysis pipelines.\n",
    "\n",
    "Typically AnnData/MuData objects are subsetted (or _sliced_) along the feature dimension using method `slice_granges`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "44501280-7b4a-4be1-9255-da9bd1465f33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(100, 11)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.slice_granges('chrX', 1, 1500000).shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8df383f3",
   "metadata": {},
   "source": [
    "## Intersection with a list of genomic ranges\n",
    "\n",
    "- Input: List of genomic ranges in pandas dataframe\n",
    "- Output: Subsetted AnnData object with inner overlapping with "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "270cfee9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gene_142</th>\n",
       "      <td>chrY</td>\n",
       "      <td>16941609</td>\n",
       "      <td>16942399</td>\n",
       "      <td>NM_014893_exon_4_0_chrY_16941610_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_849</th>\n",
       "      <td>chrY</td>\n",
       "      <td>155399</td>\n",
       "      <td>155536</td>\n",
       "      <td>NR_028057_exon_3_0_chrY_155400_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_660</th>\n",
       "      <td>chrY</td>\n",
       "      <td>244667</td>\n",
       "      <td>245252</td>\n",
       "      <td>NM_013239_exon_0_0_chrY_244668_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_493</th>\n",
       "      <td>chrY</td>\n",
       "      <td>1363220</td>\n",
       "      <td>1363354</td>\n",
       "      <td>NM_172246_exon_7_0_chrY_1363221_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_351</th>\n",
       "      <td>chrY</td>\n",
       "      <td>14533348</td>\n",
       "      <td>14533389</td>\n",
       "      <td>NR_033667_exon_4_0_chrY_14533349_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_423</th>\n",
       "      <td>chrX</td>\n",
       "      <td>150066994</td>\n",
       "      <td>150067289</td>\n",
       "      <td>NM_031462_exon_10_0_chrX_150066995_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_988</th>\n",
       "      <td>chrX</td>\n",
       "      <td>150565656</td>\n",
       "      <td>150565833</td>\n",
       "      <td>NM_001017980_exon_0_0_chrX_150565657_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_978</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153046670</td>\n",
       "      <td>153046801</td>\n",
       "      <td>NM_014370_exon_1_0_chrX_153046671_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_851</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153599240</td>\n",
       "      <td>153599729</td>\n",
       "      <td>NM_001456_exon_45_0_chrX_153599241_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene_167</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153991030</td>\n",
       "      <td>153991256</td>\n",
       "      <td>NM_001363_exon_0_0_chrX_153991031_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>79 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chrom      start        end                                    Name  \\\n",
       "Gene_142  chrY   16941609   16942399      NM_014893_exon_4_0_chrY_16941610_f   \n",
       "Gene_849  chrY     155399     155536        NR_028057_exon_3_0_chrY_155400_f   \n",
       "Gene_660  chrY     244667     245252        NM_013239_exon_0_0_chrY_244668_r   \n",
       "Gene_493  chrY    1363220    1363354       NM_172246_exon_7_0_chrY_1363221_f   \n",
       "Gene_351  chrY   14533348   14533389      NR_033667_exon_4_0_chrY_14533349_r   \n",
       "...        ...        ...        ...                                     ...   \n",
       "Gene_423  chrX  150066994  150067289    NM_031462_exon_10_0_chrX_150066995_r   \n",
       "Gene_988  chrX  150565656  150565833  NM_001017980_exon_0_0_chrX_150565657_f   \n",
       "Gene_978  chrX  153046670  153046801     NM_014370_exon_1_0_chrX_153046671_f   \n",
       "Gene_851  chrX  153599240  153599729    NM_001456_exon_45_0_chrX_153599241_r   \n",
       "Gene_167  chrX  153991030  153991256     NM_001363_exon_0_0_chrX_153991031_f   \n",
       "\n",
       "          Score Strand  \n",
       "Gene_142      0      +  \n",
       "Gene_849      0      +  \n",
       "Gene_660      0      -  \n",
       "Gene_493      0      +  \n",
       "Gene_351      0      -  \n",
       "...         ...    ...  \n",
       "Gene_423      0      -  \n",
       "Gene_988      0      +  \n",
       "Gene_978      0      +  \n",
       "Gene_851      0      -  \n",
       "Gene_167      0      +  \n",
       "\n",
       "[79 rows x 6 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.subset_by_overlap(gr).varm['coord']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d4b5ac27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>128674251</td>\n",
       "      <td>128674455</td>\n",
       "      <td>NM_000276_exon_0_0_chrX_128674252_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chrX</td>\n",
       "      <td>45016959</td>\n",
       "      <td>45017133</td>\n",
       "      <td>NM_176819_exon_2_0_chrX_45016960_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chrX</td>\n",
       "      <td>120108921</td>\n",
       "      <td>120109918</td>\n",
       "      <td>NM_001080137_exon_2_0_chrX_120108922_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chrX</td>\n",
       "      <td>17879217</td>\n",
       "      <td>17879457</td>\n",
       "      <td>NM_001172739_exon_2_0_chrX_17879218_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48814238</td>\n",
       "      <td>48814893</td>\n",
       "      <td>NM_017602_exon_8_0_chrX_48814239_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48367346</td>\n",
       "      <td>48367491</td>\n",
       "      <td>NM_203475_exon_0_0_chrX_48367347_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>977</th>\n",
       "      <td>chrX</td>\n",
       "      <td>18443724</td>\n",
       "      <td>18443815</td>\n",
       "      <td>NM_003159_exon_0_0_chrX_18443725_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>978</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153046670</td>\n",
       "      <td>153046801</td>\n",
       "      <td>NM_014370_exon_1_0_chrX_153046671_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>982</th>\n",
       "      <td>chrX</td>\n",
       "      <td>102884421</td>\n",
       "      <td>102884489</td>\n",
       "      <td>NM_001006639_exon_1_0_chrX_102884422_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>988</th>\n",
       "      <td>chrX</td>\n",
       "      <td>150565656</td>\n",
       "      <td>150565833</td>\n",
       "      <td>NM_001017980_exon_0_0_chrX_150565657_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>177</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end                                    Name  \\\n",
       "3    chrX  128674251  128674455     NM_000276_exon_0_0_chrX_128674252_f   \n",
       "16   chrX   45016959   45017133      NM_176819_exon_2_0_chrX_45016960_r   \n",
       "19   chrX  120108921  120109918  NM_001080137_exon_2_0_chrX_120108922_r   \n",
       "20   chrX   17879217   17879457   NM_001172739_exon_2_0_chrX_17879218_r   \n",
       "29   chrX   48814238   48814893      NM_017602_exon_8_0_chrX_48814239_r   \n",
       "..    ...        ...        ...                                     ...   \n",
       "973  chrX   48367346   48367491      NM_203475_exon_0_0_chrX_48367347_f   \n",
       "977  chrX   18443724   18443815      NM_003159_exon_0_0_chrX_18443725_f   \n",
       "978  chrX  153046670  153046801     NM_014370_exon_1_0_chrX_153046671_f   \n",
       "982  chrX  102884421  102884489  NM_001006639_exon_1_0_chrX_102884422_f   \n",
       "988  chrX  150565656  150565833  NM_001017980_exon_0_0_chrX_150565657_f   \n",
       "\n",
       "     Score Strand  coverage  \n",
       "3        0      +       204  \n",
       "16       0      -       156  \n",
       "19       0      -       848  \n",
       "20       0      -       238  \n",
       "29       0      -       626  \n",
       "..     ...    ...       ...  \n",
       "973      0      +       145  \n",
       "977      0      +        91  \n",
       "978      0      +        98  \n",
       "982      0      +        53  \n",
       "988      0      +       177  \n",
       "\n",
       "[78 rows x 7 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bf.coverage(exons, gr)[bf.coverage(exons, gr).coverage>0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b0b37cf8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>128674251</td>\n",
       "      <td>128674455</td>\n",
       "      <td>NM_000276_exon_0_0_chrX_128674252_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>731</th>\n",
       "      <td>chrX</td>\n",
       "      <td>148586564</td>\n",
       "      <td>148586884</td>\n",
       "      <td>NM_006123_exon_7_0_chrX_148586565_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>727</th>\n",
       "      <td>chrX</td>\n",
       "      <td>120070015</td>\n",
       "      <td>120071012</td>\n",
       "      <td>NM_001080138_exon_2_0_chrX_120070016_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>695</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153067622</td>\n",
       "      <td>153070355</td>\n",
       "      <td>NM_032512_exon_0_0_chrX_153067623_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>660</th>\n",
       "      <td>chrY</td>\n",
       "      <td>244667</td>\n",
       "      <td>245252</td>\n",
       "      <td>NM_013239_exon_0_0_chrY_244668_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153284647</td>\n",
       "      <td>153284779</td>\n",
       "      <td>NM_001025243_exon_10_0_chrX_153284648_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>chrX</td>\n",
       "      <td>48768822</td>\n",
       "      <td>48769235</td>\n",
       "      <td>NM_001042498_exon_3_0_chrX_48768823_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>chrY</td>\n",
       "      <td>15591393</td>\n",
       "      <td>15592550</td>\n",
       "      <td>NR_047610_exon_27_0_chrY_15591394_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>988</th>\n",
       "      <td>chrX</td>\n",
       "      <td>150565656</td>\n",
       "      <td>150565833</td>\n",
       "      <td>NM_001017980_exon_0_0_chrX_150565657_f</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>939</th>\n",
       "      <td>chrX</td>\n",
       "      <td>34147868</td>\n",
       "      <td>34150447</td>\n",
       "      <td>NM_203408_exon_0_0_chrX_34147869_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>78 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end                                     Name  \\\n",
       "3    chrX  128674251  128674455      NM_000276_exon_0_0_chrX_128674252_f   \n",
       "731  chrX  148586564  148586884      NM_006123_exon_7_0_chrX_148586565_r   \n",
       "727  chrX  120070015  120071012   NM_001080138_exon_2_0_chrX_120070016_r   \n",
       "695  chrX  153067622  153070355      NM_032512_exon_0_0_chrX_153067623_r   \n",
       "660  chrY     244667     245252         NM_013239_exon_0_0_chrY_244668_r   \n",
       "..    ...        ...        ...                                      ...   \n",
       "190  chrX  153284647  153284779  NM_001025243_exon_10_0_chrX_153284648_r   \n",
       "174  chrX   48768822   48769235    NM_001042498_exon_3_0_chrX_48768823_r   \n",
       "320  chrY   15591393   15592550      NR_047610_exon_27_0_chrY_15591394_r   \n",
       "988  chrX  150565656  150565833   NM_001017980_exon_0_0_chrX_150565657_f   \n",
       "939  chrX   34147868   34150447       NM_203408_exon_0_0_chrX_34147869_r   \n",
       "\n",
       "     Score Strand  count  \n",
       "3        0      +      1  \n",
       "731      0      -      1  \n",
       "727      0      -      1  \n",
       "695      0      -      1  \n",
       "660      0      -      1  \n",
       "..     ...    ...    ...  \n",
       "190      0      -      1  \n",
       "174      0      -      1  \n",
       "320      0      -      1  \n",
       "988      0      +      1  \n",
       "939      0      -      2  \n",
       "\n",
       "[78 rows x 7 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bf.count_overlaps(exons, gr)[bf.count_overlaps(exons, gr)['count']>0].sort_values('count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "491ba8eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Name</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>chrom_</th>\n",
       "      <th>start_</th>\n",
       "      <th>end_</th>\n",
       "      <th>CpG_</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chrX</td>\n",
       "      <td>34147868</td>\n",
       "      <td>34150447</td>\n",
       "      <td>NM_203408_exon_0_0_chrX_34147869_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX</td>\n",
       "      <td>34148740</td>\n",
       "      <td>34148957</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chrX</td>\n",
       "      <td>34147868</td>\n",
       "      <td>34150447</td>\n",
       "      <td>NM_203408_exon_0_0_chrX_34147869_r</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX</td>\n",
       "      <td>34149264</td>\n",
       "      <td>34149486</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom     start       end                                Name  Score  \\\n",
       "11  chrX  34147868  34150447  NM_203408_exon_0_0_chrX_34147869_r      0   \n",
       "12  chrX  34147868  34150447  NM_203408_exon_0_0_chrX_34147869_r      0   \n",
       "\n",
       "   Strand chrom_    start_      end_  CpG_  \n",
       "11      -   chrX  34148740  34148957    25  \n",
       "12      -   chrX  34149264  34149486    20  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bf.overlap(exons, gr, how='inner')[bf.overlap(exons, gr, how='inner')['start']== 34147868]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "c808b059",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAxoAAAAqCAYAAAA00i1iAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAEkElEQVR4nO3cTUhUawCA4XfsUupQIWGQZjS06AeJFlKraFURaDibwGVEY0GDFERFqDmLXJZDEQ4t3AQFLYo21q6VCANBSCC0KEIRZuHPoujHOXfV0Fzr3oWfjtd5n9053znn+85mhpdzOLEoiiIkSZIkKaCaSi9AkiRJ0vpjaEiSJEkKztCQJEmSFJyhIUmSJCk4Q0OSJElScIaGJEmSpOAMDUmSJEnBGRqSJEmSgjM0JEmSJAVnaEiSABgZGSEWi5HP5yu9FEnSOmBoSJKCeffuHRs3buTs2bNLxubm5tixYwdHjhyhWCxWYHWSpNVkaEiSgjlw4ABXr15lZGSE169fl41dv36dQqHA8PAwNTX+/UjSeucvvSRp2aIo4suXLwD09vayZ88euru7+fbtGwBjY2Pkcjl6eno4dOhQBVcqSVothoYkVZGpqSnOnTtHU1MTmzZtIpFIcPHixVIQAHz9+pUrV67Q2NhIPB4nmUxSKBTKrrN7927a29t5+fIlbW1t1NXVMTw8DEBtbS0PHjxgcnKSwcFBvn//TiqVoqWlhUwms6r3K0mqnL8qvQBJ0uqYnp7m8OHDzM3NkUql2LdvH1NTUzx9+pTPnz+Xjkun0zQ0NNDf38+HDx+4e/culy5d4smTJ2XXm5ycpKuri+7ubs6fP8/evXtLY8ePH6erq4vBwUGmp6eZmJjg+fPnxOPxVbtfSVJlGRqSVCVu3LjBzMwM4+PjtLW1lfZnMhmiKCptb9u2jVevXhGLxQAoFotks1nm5+fZunVr6bj3798zOjrKyZMnfzvfnTt3GB0dJZfL0dnZyenTp1foziRJa5GvTklSFSgWizx79oyOjo6yyPjpZ1QApFKpsu2jR4+yuLjIx48fy85JJBJ/jAyA+vp66uvrAThx4sRyb0GS9D9jaEhSFSgUCiwsLNDa2vqfx+7atatsu6GhAYDZ2dmy/YlE4l+vc/PmTWZmZti/fz/9/f1LzpckrW+GhiSpzIYNG367/9fXqwDq6ur+eI18Ps/9+/dJp9M8fvyY2dlZrl27FnSdkqS1zdCQpCrQ2NjIli1bmJiYWPG5FhcXSaVSNDU1kclkOHjwID09PTx8+JCxsbEVn1+StDYYGpJUBWpqaujs7OTFixfk8/kl4/98WrEc2WyWN2/ekM1m2bx5MwADAwPs3LmTCxcu8OPHj2BzSZLWLkNDkqrE7du32b59O8eOHePy5cvkcjkGBgZobW1lfn4+yByfPn2ir6+Pjo4OkslkaX88HmdoaIi3b98yNDQUZC5J0trm520lqUo0NzczPj5Ob28vjx49YmFhgebmZk6dOlX6OtRypdNpoiji3r17S8aSySTt7e3cunWLM2fO0NLSEmROSdLaFItCPi+XJEmSJHx1SpIkSdIKMDQkSZIkBWdoSJIkSQrO0JAkSZIUnKEhSZIkKThDQ5IkSVJwhoYkSZKk4AwNSZIkScEZGpIkSZKCMzQkSZIkBWdoSJIkSQrO0JAkSZIUnKEhSZIkKThDQ5IkSVJwhoYkSZKk4AwNSZIkScEZGpIkSZKC+xvHwSaDDv5QvAAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 1000x200 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAxoAAAAqCAYAAAA00i1iAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAD2klEQVR4nO3dv2tTexzH4XfqLVIFS9EOWtRkUtFFaAYHsZuIFSziHyBCgoNLtw4idJA6CFZwEBxcHETBQlG0c60UMurmYIeqUPxVB1GwudMt5iqXOxwSbJ5nO+d7cvI5W17h5KTUbDabAQAAKFBPpwcAAAA2HqEBAAAUTmgAAACFExoAAEDhhAYAAFA4oQEAABROaAAAAIUTGgAAQOGEBgAAUDihAcC6O3fupFQqpdFodHoUAP5wQgOAwjx//jw9PT2ZmJj47frVq1dTKpXy6NGjNk8GQLsJDQAKc+TIkdTr9Vy7di0vX75sWVtaWsrk5GTOnj2bkydPdmhCANpFaABQiGazma9fv2Zqaio7duxIvV5Ps9lcX7948WJ6e3szPT3dwSkBaBehAdBllpeXc/78+ezatSubN29OpVLJhQsX8v379/Vjvn37lvHx8QwODmbr1q0ZGxvLyspKy3nK5XJGR0fz9OnTDA8Pp6+vL7du3Up/f3+mp6fz7Nmz3L59O0ny8OHDzM7OZmpqKjt37mzr9QLQGaXmz183AbChvXnzJtVqNZ8+fUqtVsv+/fuzvLycBw8eZGFhITMzMzl37lwOHz6cgYGBjI2N5fXr17l+/XrOnDmTe/furZ+rXC6nt7c379+/T71eT7lczr59+zIyMpIkGR0dzcLCQhqNRkZGRrJ79+7Mz8+nVCp16OoBaKe/Oj0AAO0zMTGRd+/eZXFxMcPDw+v7JycnW25z2r59e+bm5tajYG1tLTdu3Mjnz5/T39+/ftyrV6/y5MmTHD9+/Jf3unnzZg4ePJhqtZovX77k8ePHIgOgi7h1CqBLrK2tZWZmJqdOnWqJjH/8HAG1Wq1l++jRo/nx40eWlpZaXlOpVH4bGUmyd+/eXL58OR8+fMj4+HgOHTpU0JUA8CcQGgBdYmVlJaurq//rA/+ePXtatgcGBpIkHz9+bNlfqVT+8zzVajVJfhs2AGxsQgOAX2zatOm3+//9s76+vr52jAPAH0hoAHSJwcHBbNu2LS9evOj0KAB0AaEB0CV6enpy+vTpzM7OptFo/LLuIYQAFMlTpwC6yJUrVzI3N5djx46lVqvlwIEDefv2be7fv5/5+flOjwfABiI0ALrI0NBQFhcXc+nSpdy9ezerq6sZGhrKiRMnsmXLlk6PB8AG4g/7AACAwvmNBgAAUDihAQAAFE5oAAAAhRMaAABA4YQGAABQOKEBAAAUTmgAAACFExoAAEDhhAYAAFA4oQEAABROaAAAAIUTGgAAQOGEBgAAUDihAQAAFE5oAAAAhRMaAABA4YQGAABQuL8Bkr/qB45DR5sAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 1000x200 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "bf.vis.plot_intervals(bf.overlap(exons, gr, how='inner'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "c605e4d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAxoAAACRCAYAAABE387UAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAHSUlEQVR4nO3cMWhU6RqA4S/GiUQHY8ZNk4AklQqBNMFOZCsRDChbWVkEI4I2dhbZoIi12G1nY0BUEG00jY0WgWAhNoKFFlEhOCZhUMxg5hYXw81dYeHeLzknmefpzslJ5hsm/w8vM2c6Wq1WKwAAABLtKHoAAABg+xEaAABAOqEBAACkExoAAEA6oQEAAKQTGgAAQDqhAQAApBMaAABAOqEBAACkExoArLl9+3Z0dHTE3Nxc0aMAsMUJDQDSzczMxPj4eAwPD0dnZ2cMDg4WPRIAm0xoAJBueno6pqeno6enJ/r7+4seB4ACCA0AUrRarfj27VtERNy4cSOWl5fjxYsXMTIyUvBkABRBaAC0mfn5+RgfH4/+/v7YtWtXDA0NxYULF2JlZWXtmu/fv8fly5ejr68v9uzZE6dPn46FhYV1f2dwcDBOnjwZT58+jdHR0eju7o6//vorIiL6+/ujUqls6vMCoFx2Fj0AAJvnw4cPceTIkVhcXIyJiYk4dOhQzM/Px/379+Pr169r1126dCl6e3tjamoq3r17Fzdv3oyLFy/G3bt31/29N2/exJkzZ+L8+fNx7ty5OHjw4GY/JQBKSmgAtJErV67Ep0+fYnZ2NkZHR9fOX7t2LVqt1trx/v37Y2ZmJjo6OiIiYnV1NW7duhVLS0vR09Ozdt3bt2/jyZMncfz48c17EgBsCT46BdAmVldX4+HDhzE2NrYuMn76GRURERMTE+uOjx49Gj9+/Ij379+v+52hoSGRAcAvCQ2ANrGwsBDLy8sxPDz8j9ceOHBg3XFvb29ERHz58mXd+aGhobwBAdhWhAYAf9PZ2fnL8//58aqIiO7u7s0YB4AtSGgAtIm+vr7Yu3dvvH79uuhRAGgDQgOgTezYsSNOnToVjx8/jrm5ub/9/L/frQCA/4dvnQJoIzdu3IiZmZk4duxYTExMxOHDh+Pjx49x7969eP78edrjvHr1Kh49ehQR//5mqqWlpbh+/XpERIyMjMTY2FjaYwFQTkIDoI0MDAzE7OxsTE5Oxp07d2J5eTkGBgbixIkTsXv37rTHefnyZUxOTq479/P47NmzQgOgDXS0vFcOAAAkc48GAACQTmgAAADphAYAAJBOaAAAAOmEBgAAkE5oAAAA6YQGAACQTmgAAADphAYAAJBOaAAAAOmEBgAAkE5oAAAA6YQGAACQTmgAAADphAYAAJBOaAAAAOl2Fj0AFKVer0ej0Sh6DKBA1Wo1arVa0WMAbEtCg7ZUr9fj6tVr0WyuFD0KUKBKpSumpv4UGwAbQGjQlhqNRjSbK7H/9z+isq+v6HGAAjQXF+LzswfRaDSEBsAGEBq0tcq+vuj6rb/oMQAAth03gwMAAOmEBgAAkE5oAAAA6YQGAACQTmgAAADphAYAAJBOaAAAAOmEBgAAkE5oAAAA6YQGAACQTmgAAADphAYAAJBOaAAAAOmEBgAAkE5oAAAA6YQGAACQTmgAAADphAYAAJBOaAAAAOmEBgAAkE5oAAAA6YQGAACQTmgAAADphAYAAJBOaAAAAOmEBgAAkE5oAAAA6XYWPQAUqbm4UPQIQEGsf4CNJTRoS9VqNSqVrvj87EHRowAFqlS6olqtFj0GwLbU0Wq1WkUPAUWo1+vRaDSKHgMoULVajVqtVvQYANuS0AAAANK5GRwAAEgnNAAAgHRCAwAASCc0AACAdEIDAABIJzQAAIB0QgMAAEgnNAAAgHRCAwAASCc0AACAdEIDAABIJzQAAIB0QgMAAEgnNAAAgHRCAwAASCc0AACAdEIDAABIt7PoAba7er0ejUaj6DFgS6lWq1Gr1YoegzZgjy43e0F5WTvlVaZ1IzQ2UL1ej6tXr0WzuVL0KLClVCpdMTX1Z2k2SrYne3T52QvKydoptzKtG6GxgRqNRjSbK7H/9z+isq+v6HFgS2guLsTnZw+i0WiUYpNk+7JHl5u9oLysnfIq27oRGpugsq8vun7rL3oMAH7BHg3/G2uHf+JmcAAAIJ3QAAAA0gkNAAAgndAAAADSCQ0AACCd0AAAANIJDQAAIJ3QAAAA0gkNAAAgndAAAADSCQ0AACCd0AAAANIJDQAAIJ3QAAAA0gkNAAAgndAAAADSCQ0AACCd0AAAANIJDQAAIJ3QAAAA0gkNAAAgndAAAADSCQ0AACCd0AAAANIJDQAAIJ3QAAAA0gkNAAAg3c6iB2gHzcWFokeALcN6YbP5nysnr0v5eY3Kp2yvidDYQNVqNSqVrvj87EHRo8CWUql0RbVaLXoMtjl7dPnZC8rJ2im3Mq2bjlar1Sp6iO2sXq9Ho9EoegzYUqrVatRqtaLHoA3Yo8vNXlBe1k55lWndCA0AACCdm8EBAIB0QgMAAEgnNAAAgHRCAwAASCc0AACAdEIDAABIJzQAAIB0QgMAAEgnNAAAgHRCAwAASCc0AACAdEIDAABIJzQAAIB0QgMAAEgnNAAAgHRCAwAASCc0AACAdP8CJEOMO8DOrboAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 1000x200 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "bf.vis.plot_intervals(df1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "24bc38a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1 = pd.DataFrame([\n",
    "    ['chr1', 1, 5],\n",
    "    ['chr1', 3, 8],\n",
    "    ['chr1', 8, 10],\n",
    "    ['chr1', 12, 14]],\n",
    "    columns=['chrom', 'start', 'end']\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "98a031f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = bf.from_any(\n",
    "    [['chr1', 4, 8],\n",
    "     ['chr1', 10, 11]], \n",
    "    name_col='chrom')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b2f9705a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  start  end  coverage\n",
       "0  chr1      1    5         1\n",
       "1  chr1      3    8         4\n",
       "2  chr1      8   10         0\n",
       "3  chr1     12   14         0"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bf.coverage(df1, df2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23d8a28a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
